docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,358 @@
+
+1. A node _must_ support at most one data companion.
+
+2. All or part of the following data _must_ be obtainable by the companion, and
+   as close to real-time as possible:
+   1. Committed block data
+   2. `FinalizeBlockResponse` data, but only for committed blocks
+
+3. The companion _must_ be able to establish the earliest height for which the
+   node has all of the requisite data.
+
+4. The API _must_ be (or be able to be) appropriately shielded from untrusted
+   consumers and abuse. Critical control facets of the API (e.g. those that
+   influence the node's pruning mechanisms) _must_ be implemented in such a way
+   as to eliminate the possibility of accidentally exposing those endpoints to
+   the public internet unprotected.
+
+5. The node _must_ know, by way of signals from the companion, which heights'
+   associated data are safe to automatically prune.
+
+6. The companion _must_ be able to handle the possibility that a node might
+   start from a non-zero height (i.e. that the node may state sync from a
+   specific height beyond genesis).
+
+7. The companion _must_ be able to handle the possibility that a node may
+   disable and then later re-enable the companion interface, potentially causing
+   the companion to have missing data in between those two heights.
+
+8. The API _must_ be opt-in. When off or not in use, it _should_ have no impact
+   on system performance.
+
+9. It _must_ not cause back-pressure into consensus.
+
+10. It _must_ not cause unbounded memory growth.
+
+11. It _must_ provide one or more ways for operators to control storage growth.
+
+12. It _must_ provide insight to operators (e.g. by way of logs/metrics) to
+    assist in dealing with possible failure modes.
+
+13. The solution _should_ be able to be backported to older versions of
+    Tendermint (e.g. v0.34).
+
+### Entity Relationships
+
+The following model shows the proposed relationships between Tendermint, a
+socket-based ABCI application, and the proposed data companion service.
+
+```
+     +----------+      +------------+      +----------------+
+     | ABCI App | <--- | Tendermint | <--- | Data Companion |
+     +----------+      +------------+      +----------------+
+```
+
+In this diagram, it is evident that Tendermint connects out to the ABCI
+application, and the companion connects to the Tendermint node.
+
+### Pruning Behaviour
+
+There are two "modes" of pruning that need to be supported by the node:
+
+1. **Data companion disabled**: Here, the node prunes as per normal based on the
+   `retain_height` parameter supplied by the application via the
+   [`Commit`][abci-commit] ABCI response.
+2. **Data companion enabled**: Here, the node prunes _blocks_ based on the lower
+   of the retain heights specified by the application and the data companion,
+   and the node prunes _block results_ based on the retain height set by the
+   data companion.
+
+Enabling the `discard_abci_responses` flag under the `[storage]` section in the
```

Comment:
I think we should go on and deprecate `discard_abci_responses` as part of this

Reply:
You don't think that some folks may want to run a full node that doesn't have a companion attached to it _and_ also doesn't store ABCI responses?

Reply:
I hope there's a decoupling between consensus layer and rpc layer, the consensus layer only need to store the raw blocks and some light weight consensus states, the ABCI responses and tx indexers belongs to the rpc layer.